### PR TITLE
FMOD Support Added - Audio Is In!

### PR DESCRIPTION
**Pull request by joncatanio**
Alright bros! Adding FMOD support was pretty fuckin' annoying so bear with me, it's also 3:00AM right now so here goes nothing. Also this might be one of those things we talk about in class so let me know if that would be easier for you all.

We now have a soundtrack when the level loads. It is three simple songs from my soundcloud lol, they work and cycle correctly so that's all that matters. This is specified in the JSON and you can add a bunch. Right now our sound cap is 512 virtual channels, I don't suspect we will need more than that but if we run into that problem I'll know what to do.

We also have sound effects yay! I was debating on doing this for this milestone but I said fuck it let's do it live. The only example right now is the fire hydrant, but the support is there for any object. To add an effect simply add it to the JSON (I'll try to update the docs on this at some point) then wherever you want to use the effect call `playEffect` (from the `AudioManager`) with the name of the effect and it'll play it for yuh.

Sound is in boys, it should be pretty straight forward from here and I am also not lagging AT ALL with the new audio additions so it's possible the hacky proof of concept was using wayy too many resources or something was in the game loop that shouldn't have been.

If you have any questions let me know. Also P.S. fuck FMOD.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,14 +55,28 @@ else()
...
+endif()
+include_directories(${FMOD_DIR}/api/lowlevel/inc)
+if(WIN32)
+   target_link_libraries(${CMAKE_PROJECT_NAME} ${FMOD_DIR}/api/lowlevel/lib/libfmod.a)
+else(APPLE)
+   target_link_libraries(${CMAKE_PROJECT_NAME} ${FMOD_DIR}/api/lowlevel/lib/libfmod.dylib)
+else()
+   target_link_libraries(${CMAKE_PROJECT_NAME} ${FMOD_DIR}/api/lowlevel/lib/libfmod.a)
+endif()
+
```
Comment (rgarmsen2295) [inline CMakeLists.txt:71]: I'm so sorry you had to do this :(

File: levels/grandmas-awakening.json [hunk 2]
```diff
@@ -348,6 +348,26 @@
...
+         "name": "HairInTheDrain",
+         "filename": "HairInTheDrain.mp3"
+      }
+   ],
+   "soundeffects": [
+      {
+         "name": "FireHydrant Clank",
+         "filename": "firehydrantclank.mp3"
+      }
+   ],
```
Comment (rgarmsen2295) [inline levels/grandmas-awakening.json:390]: Awesome

File: src/AudioManager.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+void AudioManager::update() {
+   FMOD_RESULT result;
+   system_->update();
+
+   if (stChannel_) {
+      bool playing;
+
+      /* If the current song is done get the next one. I don't believe there
+       * is a function to check if a channel is valid, so the only way to check
+       * is to call a function and check the result lol... */
```
Comment (rgarmsen2295) [inline src/AudioManager.cpp:64]: ....good lord lol

File: src/AudioManager.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+void AudioManager::addTrack(std::string filename) {
+   FMOD_RESULT result;
+   FMOD::Sound *sound = NULL;
+
+   /* `createStream` should be used for any compressed files (mp3/wav) as it
+    * will buffer the file in chunks whereas `createSound` loads the entire
+    * file into memory and uncompresses it. That is fine for simple sounds. */
+   result = system_->createStream((audioPath_ + filename).c_str(),
+      FMOD_LOOP_NORMAL | FMOD_2D, 0, &sound);
```
Comment (rgarmsen2295) [inline src/AudioManager.cpp:89]: Nice, I imagine this might have fixed the perf issues as well

**Review by rgarmsen2295 — APPROVED** [review]
Looks great - awesome, awesome stuff, I know this was your 12 Herculean tasks all in one.

File: CMakeLists.txt [hunk 5]
```diff
@@ -55,14 +55,28 @@ else()
...
 # OS specific options and libraries
 if(WIN32)
   # c++0x is enabled by default.
   # -Wall produces way too many warnings.
   # -pedantic is not supported.
 else()
   # Enable all pedantic warnings.
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x -Wall -pedantic -DDEBUG_BB")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x -Wall -pedantic")
```
Comment (nurgan) [inline CMakeLists.txt]: thx, I was sick of changing it all the time

**Review by nurgan — APPROVED** [review]
Looks good to me. Well, fmod actually looks like shit to me, but your code looks good xD

**Pull request merged**